Source/ParticleInterpolator.H

```diff
@@ -0,0 +1,133 @@
+#ifndef PARTICLE_INTERPOLATOR_H_
+#define PARTICLE_INTERPOLATOR_H_
+
+/*
+   The ParticleInterpolator provides cell indexing and shape
+   functions for Particle-in-Cell spline interpolation between
+   particles and an underlying regularly spaced grid.
+
+   Template specializations are provided for spline orders 0, 1, and 2.
+
+   Cell indexing is only defined for cell centered grid data,
+   and it provides cell indices relative to the leftmost cell in the domain.
+   That is, if the leftmost interior cell is index N, you will need
+   to offset the indices returned by this interpolator by +N.
+
+   This interpolator works with coordinates in units of the grid cell spacing
+   using a variable delta defined as follows:
+       delta = (x_particle - xlo)/dx
+   where:
+       * x_particle: the particle coordinate
+       * xlo: the coordinate of the left (lo) edge of the domain
+       * dx: the cell spacing
+
+   The API provided by ParticleInterpolator is:
+       * first(): returns the first cell index supporting the particle shape
+       * last(): returns the last cell index supporting the particle shape
+       * operator(int i): returns the shape function for cell index `first() <= i <= last()`
+*/
+
+#include <AMReX_REAL.H>
+
+using namespace amrex;
+
+template <int max_spline_order>
+struct ParticleInterpolator
+{
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    ParticleInterpolator(Real delta, int order) {
+        // although we've set the max order at compile time we can
+        // always choose to run with a lower order at runtime.
+        if (max_spline_order >= 2 && order > 1)
+            compute_order_2(delta);
+        else if (max_spline_order >= 1 && order > 0)
+            compute_order_1(delta);
+        else
+            compute_order_0(delta);
+    }
+
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    const Real& operator() (int i) const noexcept {
```

Comment:
Could you remind me why these return references rather than directly returning by value?

Reply:
The only thing I can think of is if we wanted to access/modify the elements themselves. But since this is a const function, I don't think that is a use case.

Reply:
The reference makes sure the compiler does not make a copy. It would hopefully realize not to when returning by value, but we don't have to rely on that.

Reply:
right, that's not a use case ... I would assert that no user should be able to modify the first index, last index, or shape factors themselves. I can't think of any reason why that would be necessary and it seems like it could invite bugs.